packages/prisma-client/prisma/migrations/20220909124542_move_to_builds/migration.ts

```diff
@@ -0,0 +1,44 @@
+import { PrismaClient, Prisma } from "./client";
+import { v4 as uuidv4 } from "uuid";
+
+export default () => {
+  const client = new PrismaClient();
+  return client.$transaction(async (prisma) => {
+    const trees = await prisma.tree.findMany();
+    const projects = await prisma.project.findMany();
+
+    const builds: Prisma.BuildCreateManyInput[] = [];
+
+    for (const tree of trees) {
+      const project = projects.find(
+        (project) =>
+          project.devTreeId === tree.id ||
+          project.prodTreeId === tree.id ||
+          JSON.parse(project.prodTreeIdHistory).includes(tree.id)
+      );
+
+      if (project === undefined) {
+        continue;
+      }
+
+      builds.push({
+        pages: JSON.stringify({
+          homePage: {
+            id: uuidv4(),
+            name: "Home",
+            path: "",
+            title: "Home",
+            meta: {},
+            treeId: tree.id,
+          },
+          pages: [],
+        }),
+        isDev: project.devTreeId === tree.id,
+        isProd: project.prodTreeId === tree.id,
+        projectId: project.id,
+      });
+    }
+
+    await prisma.build.createMany({ data: builds });
```

Comment:
did you mean to run this synchronously? or could all of these run in parallel?

Reply:
Not sure I understand the question. If you're asking why I use `createMany` instead of `create`, that's because with `createMany` Prisma should make just one SQL query (it may split in to several batches, but still less that a query for each row)

Reply:
Doesn't matter for the amount of data we a're working with. Just want to set a good example.

Reply:
what I mean is  that this current implementation will run one query after another in that for of loop, but it could run in parallel, I see 2 options:

1. using map, return promises from queries without awaiting, then Promise.all - this will send many queries all in parallel
2. use transaction API https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api, not sure if prisma does there anything different than in 1 (I recall they can optimize that, not sure though)

Reply:
> one query after another in that "for of" loop

There're no queries in the loop. Entire migration is 3 queries.

> using map, return promises from queries without awaiting, then Promise.all - this will send many queries all in parallel

This will work only if Prisma uses several connections to the database AND if we won't wrap the entire migration into a transaction (a transaction can't be run over multiple connections afaik)

> use transaction API https://www.prisma.io/docs/concepts/components/prisma-client/transactions#the-transaction-api, not sure if prisma does there anything different than in 1 (I recall they can optimize that, not sure though)

Same. Prisma can't run queries in a transaction in parallel, that's just a limitation on Postgres level

Reply:
Either I am missing it or you are: we are running those queries from some server. That server has a latency to the database server.

When we send the query and await for a response, we are waiting for execution + latency to receive the results.
When we parallelize like I said - queries are send all at once, we are waiting for all responses, no matter which one comes back first.

So I am optimizing away the latency between the database server and the node server.

Reply:
If it's just 3 queries it may not matter, but in a complex migration that could be many queries

Reply:
you are right, I misread when this query happens because of indentations, I thought it's inside the for-of loop but its after that and you are already using transaction, so I guess prisma will send all queries together